locator/abstract_replication_strategy.cc

```diff
@@ -241,7 +241,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
             dht::token_range_vector r = tm.get_primary_ranges_for(t);
             rslogger.debug("token={} primary_range={} endpoint={}", t, r, endpoint);
             for (auto&& rng : r) {
-                ret.emplace(ep, rng);
+                ret.push_back(rng);
```

Comment:
The change looks reasonable, but I don't see how the existing code caused a problem. How many address ranges can an endpoint participate in? At most 256*rf.

Unless it's everywhere_replication_strategy, in which case there are 256*nodes vnodes that are relevant to an endpoint. Still, even that doesn't explain how the multimap blew up.

Reply:
@avikivity it looks like the multimap blows up since we don't use an optimized structure for the token range.

I ran a small sanity test as follows and it it shows it's in the ball park:

```c++
#include <unordered_map>
#include <seastar/net/inet_address.hh>
#include <seastar/core/memory.hh>

SEASTAR_TEST_CASE(test_unordered_multimap) {
    enum class token_kind {
        before_all_keys,
        key,
        after_all_keys,
    };
    struct token {
        token_kind kind;
        int64_t data;
    };
    struct interval_bound {
        token value;
        bool inclusive;
    };
    struct token_range {
        std::optional<interval_bound> start;
        std::optional<interval_bound> end;
        bool singular;
    };
    std::unordered_multimap<net::inet_address, token_range> mm;
    int64_t tokens = 256;
    int nodes = 50;

    auto free_before = memory::stats().free_memory();

    for (auto node = 0; node < nodes; node++) {
        auto ia = net::inet_address(fmt::format("127.0.0.{}", node));
        for (int64_t t = 0; t < tokens; t++) {
            mm.emplace(ia, token_range{
                    interval_bound{token{token_kind::key, t}, true},
                    interval_bound{token{token_kind::key, t + 1}, false},
                    true});
        }
    }

    auto free_after = memory::stats().free_memory();

    fmt::print("consumed {} bytes", free_before - free_after);
    return make_ready_future<>();
}
```

```
Running 1 test case...
WARN  2021-12-05 11:40:35,498 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
random-seed=2728524117
consumed 3031040 bytes
```

Reply:
Sorry, this doesn't necessarily explains the size of the "buckets" vector, but the overall memory consumption.

Reply:
Moving discussion to https://github.com/scylladb/scylla/issues/9667